apollo-router/src/plugins/rhai.rs

```diff
@@ -0,0 +1,775 @@
+use std::{path::PathBuf, str::FromStr, sync::Arc};
+
+use apollo_router_core::{
+    plugin_utils, Context, ExecutionRequest, ExecutionResponse, QueryPlannerRequest,
+    QueryPlannerResponse, SubgraphRequest, SubgraphResponse,
+};
+use apollo_router_core::{
+    register_plugin, Error, Object, Plugin, RouterRequest, RouterResponse, Value,
+};
+use futures::executor::block_on;
+use http::HeaderMap;
+use http::{header::HeaderName, HeaderValue};
+use rhai::serde::{from_dynamic, to_dynamic};
+use rhai::{Dynamic, Engine, EvalAltResult, Scope, AST};
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde_json_bytes::ByteString;
+
+use tower::{util::BoxService, BoxError, ServiceExt};
+
+const CONTEXT_ERROR: &str = "__rhai_error";
+
+#[derive(Default, Clone)]
+struct Rhai {
+    ast: AST,
+}
+
+#[derive(Deserialize, JsonSchema)]
+struct Conf {
+    filename: PathBuf,
+}
+
+trait RhaiObjectSetterGetter {
+    fn set(&mut self, key: String, value: Value);
+    fn get_cloned(&mut self, key: String) -> Value;
+}
+
+impl RhaiObjectSetterGetter for Object {
+    fn set(&mut self, key: String, value: Value) {
+        self.insert(ByteString::from(key), value);
+    }
+    fn get_cloned(&mut self, key: String) -> Value {
+        self.get(&ByteString::from(key))
+            .cloned()
+            .unwrap_or_default()
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Headers(HeaderMap);
+impl Headers {
+    fn set_header(&mut self, name: String, value: String) -> Result<(), Box<EvalAltResult>> {
+        self.0.insert(
+            HeaderName::from_str(&name)
+                .map_err(|err| format!("invalid header name '{name}': {err}"))?,
+            HeaderValue::from_str(&value)
+                .map_err(|err| format!("invalid header value '{value}': {err}"))?,
+        );
+        Ok(())
+    }
+    fn get_header(&mut self, name: String) -> String {
+        self.0
+            .get(&name)
+            .cloned()
+            .map(|h| Some(h.to_str().ok()?.to_string()))
+            .flatten()
+            .unwrap_or_default()
+    }
+}
+
+impl Plugin for Rhai {
```

Comment:
A lot of the code in the various *_service functions is "boilerplate". I can tell you've done a good job standardising the code, but could we go a bit further and have some macros to generate this code?

I imagine two macros:
service_handle_request!()
service_handle_response!()

I think. they probably only need one parameter which is the FUNCTION_NAME_* string...

Reply:
Thanks for your suggestion ! I implemented a macro for the response part, the request part is often different because we sometimes have an Arc, sometimes not, and we're doing different things for each kind of requests. It becomes cursed and not very maintainable if I try to put everything in a macro for this part.